### DISCUSSION TITLE
Configurable `__manifest` route for `lazyRouteDiscovery` feature

### DISCUSSION THREAD

**Discussion by johannesbraeunig** [remix-run/react-router#12418] [Proposals]
## **Background**  

We're running **multiple Remix applications on a single domain**, structured by distinct `basePath` segments. For example:  
- `http://remix.run.com/register` → **Register App**  
- `http://remix.run.com/docs` → **Docs App**  

Each app operates independently and can have deeply nested routes, such as:  
`http://remix.run.com/docs/intro/first/page/` → served by the **Docs App**.

---

## **The Issue**  

After enabling the future flag `v3_lazyRouteDiscovery` in a Remix app, we encountered a critical problem:  

**The `__manifest` route is shared across all apps on the same domain.**  
This behavior assumes the `__manifest` route exists at the root level of the domain, causing conflicts and breaking the app where this flag was enabled.  

This issue has been previously highlighted by @JLarky: [Discussion Comment](https://github.com/remix-run/react-router/discussions/11113#discussioncomment-10068924).

---

## **Proposed Solution**  

To resolve this, the destination of the `__manifest` route should be **configurable** or **disableable**.
As mentioned by @brophdawg11 in the linked discussion, it seems this functionality may already exist but has not been shipped yet.

By allowing customization of the `__manifest` route location, RR7/Remix applications can operate independently on the same domain without conflict, ensuring compatibility with the `lazyRouteDiscovery` feature.

---

We’d appreciate feedback on whether this configuration is already available or if there are plans to introduce it soon.

Thank you for your work and support! 🙌

**Comment by philippd** [comment]
We’ve encountered the same issue and would also like the `__manifest` route to be configurable.

An even better solution for us, given our relatively small manifest file, would be a configuration option to opt out of lazy route discovery entirely and keep serving a static manifest file instead.

**Comment by johannesbraeunig** [comment]
@brookslybrand do you think you can take a look at this request? If you need more context, please don't hesitate to let me know! Thank you in advance 🙌

**Comment by KnisterPeter** [comment]
Another solution would be that the `__manifest` route respects the `basePath` setting of vite. I guess that's the most expected way to solve it.

**Reply by JLarky** [reply]
I hear the same argument used by Astro team (which has very similar global magic routes) but that makes deployment story less flexible. I.e. that means that I can only have one global app that uses urls without a prefix and everyone else have to be scoped to `/dashboard`, `/app` and so on. I mean it's definatelly better than nothing, but in case of Remix they don't have as many magical urls yet and opportinity to make them configurable is easier to achieve :)

**Reply by KnisterPeter** [reply]
I meant when there is a base path configured in vite, I would have expected that all path (!) respect the base path.
That does not mean that configuration is not a good thing and would be good as well.

**Reply by KnisterPeter** [reply]
What I proposed is already possible and works as expected.
There is a `basename` config in the remix vite plugin config. It's respected by the router for lazy route discovery.

**Reply by johannesbraeunig** [reply]
Thank you for the investigations! ♥️ This could be improved in the documentation. I take a look and contribute it. 🤜🤛

I'll close this discussion.

**Comment by s-abstract** [comment]
In my case, the Lazy Route Discovery feature breaks PWA in offline mode. And I cannot complete Remix -> React Router migration because of that. That would be helpful to be able to opt out of this feature, unless you have a better solution for PWA and Lazy Route Discovery.

**Comment by fernandojbf** [comment]
I’m currently patching `react-router` to address the issue you mentioned. I’m facing the exact same problem. Using `basename` is not an option since it impacts all routes of the app.

**Reply by JLarky** [reply]
yeah, basename is only an option if your app is not at the root of the domain like /app1/ and /app2/ but if you have one app handling /login /logout and /dashboard and the other app that handles /blog and /news then you are out of luck

**Reply by fernandojbf** [reply]
For projects using application slicing via different applications that does not have a clear root basename this can be a problem...

**Comment by joshuaellis** [comment]
I'd be interested in this endpoint being configurable as well, for instance at sanity we have multiple apps on the same domain, where the assets are namespaced to avoid collisions, the app's don't have basename separators e.g. `/dashboard` `/manage`. So ideally i'd change `__manifest` to be `/dashboard/__manifest` – doesn't look like this is possible however?

**Comment by nobleach** [comment]
I agree to the above, After upgrading from Remix, and a subsequent deploy to our dev/staging/qa environments, this problem reared its heads. I do not have the opiton of rewiring all my routes with a different basename at the moment. Even if I did, there are other problems I'd have to face (like the first component in my routes being dynamic `/us`, `/uk` etc - so setting ONE basename isn't an option). Is there a reason that making the `__manifest` route configurable isn't an option?

**Comment by philippd** [comment]
fyi: A new `routeDiscovery` config has been added in https://github.com/remix-run/react-router/pull/13451 that will give you the ability to configure the manifest path. This will be available in the next release.

**Reply by nobleach** [reply]
That is such fantastic news. I was getting ready to prepare a great argument on why we cannot simply use the existing workaround. I appreciate this heads up!

**Comment by fernandojbf** [comment]
`routeDiscovery` just got released in [v7.6.0](https://github.com/remix-run/react-router/blob/main/CHANGELOG.md#routediscovery-config-option). Thanks
